examples/checker/main.go

```diff
@@ -0,0 +1,196 @@
+// Copyright 2017-2019 the u-root Authors. All rights reserved
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// Checker is a troubleshooting and auto-remediation tool to help you identify
+// issues that won't let your system boot.
+// It will simply run through a checklist of checks and optional remediations,
+// and continue until it either succeeds or fails without recovery.
+//
+// The checklist is specified via configuration files in JSON format. Run the
+// `genconfig` subcommand to get an example configuration file.
+package main
+
+import (
+	"bytes"
+	"errors"
+	"flag"
+	"fmt"
+	"html/template"
+	"io/ioutil"
+	"log"
+	"os"
+	"strings"
+
+	"github.com/u-root/u-root/pkg/checker"
+)
+
+var (
+	flagsetRun    = flag.NewFlagSet("run", flag.ExitOnError)
+	flagRunConfig = flagsetRun.String("c", "", "Configuration file")
+	flagRunParams = flagsetRun.String("p", "", "Template parameters, comma-separated key=value, to use text/template in config files. Example: interface=eth0,linkspeed=40G")
+
+	flagsetGenconfig = flag.NewFlagSet("genconfig", flag.ExitOnError)
+)
+
+// CurrentVersion specifies the current version for the configuration format.
+const CurrentVersion = "0.1"
+
+var sampleConfig = fmt.Sprintf(`{
+    "version": "%s",
+	"checks": [
+		{
+			"check": {
+				"name": "printf",
+				"description": "example printf",
+				"args": ["this is an %%s", "example"]
+			},
+			"remediation": {
+				"name": "cmd",
+				"description": "just return true",
+				"args": ["/bin/true"]
+			}
+		},
+		{
+			"check": {
+				"name": "interface_exists",
+				"description": "check that interface lo exists",
+				"args": ["lo"]
+			}
+		}
+	]
+}`, CurrentVersion)
+
+var (
+	checkRunners = map[string]interface{}{
+		"interface_exists":             checker.InterfaceExists,
+		"printf":                       printfWrapper,
+		"cmd":                          checker.RunCmd,
+		"link_speed":                   checker.LinkSpeed,
+		"link_autoneg":                 checker.LinkAutoneg,
+		"interface_has_linklocal_addr": checker.InterfaceHasLinkLocalAddress,
+		"interface_has_global_addrs":   checker.InterfaceHasGlobalAddresses,
+		"interface_can_do_dhcpv6":      checker.InterfaceCanDoDHCPv6,
```

Comment:
At the very first sight, if I understand correctly, the necessity of reflection grows from here (because here we have `interface{}`). Basically if all this functions just had the same signature:
```
func(args ...string) error
```
then we could've avoid reflection:
* We could've make map `map[string]func(...string)error`
* We could've fill Check.Run with `func() error { return checkRunners[name](item.Check.Args...) }`

The essential downside in this case: we will stop checking amount of arguments on stage of `ResolveChecklist` (and will move this checks later, which is bad: more duplication and later error reporting).

Is this correct?